Remotion/Web/Development.Analyzers/RMWEB0001_WebStringAnalyzer.cs

```diff
@@ -0,0 +1,223 @@
+      {
+        var argumentType = GetFirstArgumentTypeSymbol (node, ctx.SemanticModel);
+        if (SymbolEqualityComparer.Default.Equals (argumentType, symbols.WebStringTypeSymbol)
+            || SymbolEqualityComparer.Default.Equals (argumentType, symbols.PlainTextStringTypeSymbol))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              symbols.StringBuilderAppendMethodSymbol.ToString(),
+              argumentType.ToString(),
+              "Call '.ToString(WebStringEncoding.HtmlWithTransformedLineBreaks)' on the WebString instance instead.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+      else if (SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, symbols.HtmlTextWriterWriteMethodSymbol))
+      {
+        var argumentType = GetFirstArgumentTypeSymbol (node, ctx.SemanticModel);
+        if (SymbolEqualityComparer.Default.Equals (argumentType, symbols.WebStringTypeSymbol)
+            || SymbolEqualityComparer.Default.Equals (argumentType, symbols.PlainTextStringTypeSymbol))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              symbols.HtmlTextWriterWriteMethodSymbol.ToString(),
+              argumentType.ToString(),
+              $"Use '{argumentType.ToString()}.Write(HtmlTextWriter)' instead.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+      else if (AnyWithResult (
+          symbols.StringJoinMethodSymbols,
+          s => SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, s),
+          out var methodSymbol))
+      {
+        var argumentTypes = GetArgumentTypeSymbols (node, ctx.SemanticModel);
+        if (AnyWithResult (
+            argumentTypes,
+            a => SymbolEqualityComparer.Default.Equals (a, symbols.WebStringTypeSymbol)
+                 || SymbolEqualityComparer.Default.Equals (a, symbols.PlainTextStringTypeSymbol)
+                 || ctx.Compilation.ClassifyConversion (a, symbols.WebStringEnumerable).IsImplicit
+                 || ctx.Compilation.ClassifyConversion (a, symbols.PlainTextStringEnumerable).IsImplicit,
+            out var result))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              methodSymbol.ToString(),
+              result.ToString(),
+              $"Encode the {result.ToString()} instances first.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+    }
+
+    private static bool AnyWithResult<TSymbol> (IEnumerable<TSymbol> methods, Func<TSymbol, bool> predicate, out TSymbol match)
+        where TSymbol : ISymbol
+    {
+      match = methods.FirstOrDefault (predicate)!;
+      return match != null;
```

Comment:
Stylistic question: to me, `return method.Any(m => m==predicate);` feels like it's easier to understand / more on point of the intended semantics. Good idea / different idea / something I missed?

Reply:
It would be alright, I just didn't want to iterate twice over the methods if I don't have to.

Reply:
Am confused: why do you iterate twice with Any and only once with FirstOrDefault?

Reply:
How would I assign the `out` parameter only using `Any`? The use-case for this method is to provide a way of checking and extracting the matching element inside of an if condition

Reply:
Ah, there's an out parameter. Hmm...would it be cleaner to return a nullable value instead of the out parameter and then capture the result in a boolean expression in the if. 
`if (var result = Any(...) != null)`

Am not sure if that's valid C#9, though, or just wishful thinking.

Another consideration would be to find a name that works with the Try... naming convention associated with boolean + out. Or we result a value tuple of (boolean, method), but that seems strange when we could just return method. => things I do to avoid out-parameters. Thoughts?

Reply:
I think declarations are statements but assignments expressions. So it should work when the variable is already declared - which is 😒. I'll settle for some Try variant naming.

Reply:
Okay then. 
(why the sad-face for "variable already declared"?)

And we can always brainstorm the name if you want. The Try-name is just the typical go to based on .net framework patterns.

Reply:
> why the sad-face for "variable already declared"?

I don't want to declare a variable in the method only assigned and used in one branch